algorithms.tex

```diff
@@ -8,6 +8,143 @@
+series, and ODE solutions. All features work in arbitrary precision
+and use algorithms that support computig hundreds of digits rapidly,
+except in degenerate cases.
+
+The double exponential (tanh-sinh) quadrature is used for numerical
+integration by default. For smooth integrands, this algorithm usually
+converges extremely rapidly, even when the integration interval is infinite
+or singularities are present at the endpoints.
+However, for good performance, singularities
+in the middle of the interval must be specified
+the user.
+To evaluate slowly converging limits and infinite series, mpmath
+automatically attempts to apply Richardson extrapolation and the
+Shanks transformation (Euler-Maclaurin summation can also be used).
+A function to evaluate oscillatory integrals by means of convergence
+acceleration is also available.
+
+A wide array of higher mathematical functions are implemented
+with full support for complex values of all parameters and arguments,
+including complete and incomplete gamma functions,
+Bessel functions, orthogonal polynomials, elliptic functions and integrals,
+zeta and polylogarithm functions,
+the generalized hypergeometric function, and the Meijer G-function.
+
+Most special functions are implemented as linear 
+combinations of the generalized hypergeometric function ${}_pF_q$,
+which is computed by a combination of direct summation,
+argument transformations (for ${}_2F_1$, ${}_3F_2$, $\ldots$)
+and asymptotic expansions
+(for ${}_0F_1$, ${}_1F_1$, ${}_1F_2$, ${}_2F_2$, ${}_2F_3$)
+to cover the whole complex domain.
+Numerical integration and generic convergence acceleration
+are also used in a few special cases.
+
+In general, linear combinations and argument transformations
+give rise to singularities that have to be removed for certain
+combinations of parameters.
+A good example is the modified Bessel function of the second kind
+$$K_{\nu}(z) = \frac{1}{2} \left[
+            \left(\frac{z}{2}\right)^{-\nu}
+                \Gamma(\nu)
+                {}_0F_1\left(1-\nu, \frac{z^2}{4}\right)
+             -
+             \left(\frac{z}{2}\right)^{\nu}
+                 \frac{\pi}{\nu \sin(\pi \nu) \Gamma(\nu)}
+                 {}_0F_1\left(\nu+1, \frac{z^2}{4}\right)
+            \right]$$
+where the limiting value $\lim_{\varepsilon \to 0} K_{n+\varepsilon}(z)$
+has to be computed when $\nu = n$ is an integer.
+A generic algorithm is used to evaluate
+hypergeometric-type linear combinations of the above type.
+The implementation automatically detects cancellation problems,
+and computes limits numerically by perturbing parameters whenever
+internal singularities occur (the perturbation size is automatically
+decreased until the result is detected to converge numerically).
+
+It is important for a computer algebra system
+to support generalized hypergeometric functions and
+Meijer G-functions robustly, due to their frequent appearance
+in closed forms for integrals and sums.
```

Comment:
Would be worth pointing out at this point that SymPy does in fact implement algorithms for computing integrals in terms of Meijer G-functions.